composer/trainer/trainer.py

```diff
@@ -300,32 +300,49 @@ def __init__(
         self.state.schedulers = ComposedScheduler(schedulers=schedulers)
 
-        # TODO(#121): get checkpointing working with DeepSpeed.
+        assert isinstance(self.state.model, BaseMosaicModel)
+        self.original_model = self.state.model  # type: ignore  # TODO(ravi) -- update the state to add an original model helper
+
         self.checkpoint_saver = None
-        if checkpoint_interval is not None and checkpoint_interval_unit is not None:
-            self.checkpoint_saver = CheckpointSaver(checkpoint_interval_unit=checkpoint_interval_unit,
+        if checkpoint_folder and checkpoint_interval and checkpoint_interval_unit:
+            self.checkpoint_saver = CheckpointSaver(checkpoint_folder=get_relative_to_run_directory(checkpoint_folder),
                                                     checkpoint_interval=checkpoint_interval,
-                                                    checkpoint_folder=get_relative_to_run_directory(checkpoint_folder))
-
-            if self.deepspeed_enabled:
-                raise NotImplementedError("Checkpointing is not yet supported with DeepSpeed.")
+                                                    checkpoint_interval_unit=checkpoint_interval_unit)
 
-        # TODO(#121): get checkpointing working with DeepSpeed.
         self.checkpoint_loader = None
-        if checkpoint_filepath is not None:
-            if self.deepspeed_enabled:
-                raise NotImplementedError("Checkpointing is not yet supported with DeepSpeed.")
-
+        if checkpoint_filepath:
             self.checkpoint_loader = CheckpointLoader(checkpoint_filepath=checkpoint_filepath,
                                                       load_weights_only=checkpoint_load_weights_only,
                                                       strict_model_weights=checkpoint_strict_model_weights)
 
-            restored_seed = self.checkpoint_loader.load_checkpoint(state=self.state)
-            # Set the restored seed so that the correct seed will be saved in future checkpoints
-            # Used to handle the case where another checkpoint is saved after resuming from checkpoint.
-            # In this case, self.seed is stored in the second checkpoint so it must have the correct value.
-            if restored_seed is not None:
-                self.seed = restored_seed
+        # place the state, model in the proper devices, and initialize from a checkpoint if provided
+        if self.deepspeed_enabled:
+            import deepspeed
+
+            assert self.deepspeed_hparams is not None
+            deepspeed_config = self.deepspeed_hparams.initialize_object(self.state, self.grad_clip_norm)
+            optimizer = ensure_tuple(self.state.optimizers)[0]
+            (self.state.model, self.state.optimizers, _, _) = deepspeed.initialize(
+                config=deepspeed_config,
+                model=self.state.model,
+                optimizer=optimizer,
+            )
+
+            if self.checkpoint_loader:
```

Comment:
nit: unindent this block, replace line 335 with `if not self.deepspeed_enabled:`, and remove the duplicated code on lines 336-339.